From 087538df097d352995952fb1f36cba3d1fca27c2 Mon Sep 17 00:00:00 2001
From: James Chapman <jachapma@redhat.com>
Date: Sun, 6 Apr 2025 21:23:08 +0000
Subject: [PATCH] Issue 6713 - ns-slapd crash during mdb offline import (#6714)

Bug description: A segmentation fault is triggered in
dbmdb_import_prepare_worker_entry() during an mdb offline import.

The import producer thread parses, validates and writes ldif entries
to the worker queue, while the import worker threads simultaneously read,
format and index entries before adding them to the DB. A race condition
occurs when a worker thread reads an entry before the producer has
fully written it, leading to a segmentation fault.

Fix description: Ensure thread safe access by locking the worker queue
before writing entries.

Fixes: https://github.com/389ds/389-ds-base/issues/6713

Reviewed by: @progier389, @bordaz (Thank you)
---
 ldap/servers/slapd/back-ldbm/db-mdb/mdb_import_threads.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/ldap/servers/slapd/back-ldbm/db-mdb/mdb_import_threads.c b/ldap/servers/slapd/back-ldbm/db-mdb/mdb_import_threads.c
index 0f445bb56..39d2b06f7 100644
--- a/ldap/servers/slapd/back-ldbm/db-mdb/mdb_import_threads.c
+++ b/ldap/servers/slapd/back-ldbm/db-mdb/mdb_import_threads.c
@@ -417,13 +417,14 @@ dbmdb_import_workerq_push(ImportQueue_t *q, WorkerQueueData_t *data)
             safe_cond_wait(&q->cv, &q->mutex);
         }
     }
-    pthread_mutex_unlock(&q->mutex);
     if (q->job->flags & FLAG_ABORT) {
         /* in this case worker thread does not free the data so we should do it */
         dbmdb_import_workerq_free_data(data);
+        pthread_mutex_unlock(&q->mutex);
         return -1;
     }
     dbmdb_dup_worker_slot(q, data, slot);
+    pthread_mutex_unlock(&q->mutex);
     return 0;
 }
 
-- 
2.49.0

